riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1825 @@
+    instr jump l: label, WL
+        link ~> regs.mstore(0, WL, STEP, tmp1_h, tmp1_l)
+        link => byte.check(tmp1_h)
+    {
+        pc + 1 = (tmp1_h * 2**16) + tmp1_l,
+        pc' = l
+    }
+    
+    col witness tmp_fe;
+
+    // Jump to the address in register X and store the return program counter in register W.
+    instr jump_dyn XL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, WL, STEP, tmp2_h, tmp2_l)
+        // pc is capped at 24 bits, so for this instruction 
+        // we restrict the higher limbs to 1 byte
+        link => byte.check(tmp1_h)
+        link => byte.check(tmp2_h)
+    {
+        // We need a temporary col here because the compiler doesn't like
+        // `exp` inside an expr on the RHS of a pc update. 
+        tmp_fe = (tmp1_h * 2**16) + tmp1_l,
+        pc' = tmp_fe,
+        pc + 1 = tmp2_h * 2**16 + tmp2_l
+    }
+
+    // Jump to `l` if val(X) != val(Y) is nonzero, where X and Y are register ids.
+    instr branch_if_not_equal XL, YL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link => byte.check(tmp3_h)
+        link => byte.check(tmp3_l)
+        link => byte.check(tmp4_h)
+        link => byte.check(tmp4_l)
+        link => byte.check(tmp5_h)
+        link => byte.check(tmp5_l)
+        link => byte.check(tmp6_h)
+        link => byte.check(tmp6_l)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+
+        // What we want to do is to compare tmp1_h == tmp2_h && tmp1_l == tmp2_l.
+        // This could be done with the constraint below, but overflows causes soundness issues:
+        // XX = (tmp1_h - tmp2_h)**2 + (tmp1_l - tmp2_l)**2,
+        // One solution is to further decompose into single bytes.
+
+        // Decompose tmp1_h into 2 bytes
+        tmp1_h = tmp3_h * 2**8 + tmp3_l,
+        // Decompose tmp1_l into 2 bytes
+        tmp1_l = tmp4_h * 2**8 + tmp4_l,
+        // Decompose tmp2_h into 2 bytes
+        tmp2_h = tmp5_h * 2**8 + tmp5_l,
+        // Decompose tmp2_l into 2 bytes
+        tmp2_l = tmp6_h * 2**8 + tmp6_l,
+
+        // now we have that
+        // 32bits tmp1 = tmp3_h * 2**24 + tmp3_l * 2**16 + tmp4_h * 2**8 + tmp4_l
+        // 32bits tmp2 = tmp5_h * 2**24 + tmp5_l * 2**16 + tmp6_h * 2**8 + tmp6_l
+
+        // The constraint below leads to a plonky3 translation error:
+        // thread '<unnamed>' panicked at /home/leo/devel/powdr2/plonky3/src/circuit_builder.rs:259:25:
+        // internal error: entered unreachable code: exponentiations should have been evaluated
+        // So we inline the squares.
+
+        // This should fit in 19 bits.
+        //XX = (tmp3_h - tmp5_h)**2 + (tmp3_l - tmp5_l)**2 + (tmp4_h - tmp6_h)**2 + (tmp4_l - tmp6_l)**2,
+        XX = (tmp3_h - tmp5_h)*(tmp3_h - tmp5_h) + (tmp3_l - tmp5_l)*(tmp3_l - tmp5_l) + (tmp4_h - tmp6_h)*(tmp4_h - tmp6_h) + (tmp4_l - tmp6_l)*(tmp4_l - tmp6_l),
```

Comment:
@Schaeff @georgwiese this here

Reply:
Defining the euclidean norm (squared) in PIL could make this neater.

In pseudo-PIL:

```rust
let squared_euclidean_distance = |left: fe[], right: fe[]| {
   assert_eq(len(left), len(right));
   array::fold(array::map(array::zip(left, right), |(left, right)| (left - right) * (left - right)), |acc, e| acc + e)
};

XX = squared_euclidean_distance([tmp3_h, tmp3_l, tmp5_h, tmp5_l], [tmp4_h, tmp4_l, tmp6_h, tmp6_l]),
```

Maybe this is not needed if we only use this once though.

Reply:
Comment still there and squares not there yet

Reply:
Oh weird, will check